vagrant-tools/run-tests.sh

```diff
@@ -0,0 +1,164 @@
+fi
+
+if id "vagrant" >/dev/null 2>&1; then
+  test_user="vagrant"
+elif id "ubuntu" >/dev/null 2>&1; then
+  test_user="ubuntu"
+elif [[ "$SUDO_USER" ]] && [[ "$SUDO_USER" != root ]]; then
+  test_user="$SUDO_USER"
+  >&2 echo "${bold_out}${blue_out}It looks like this is a host dev environment not a Vagrant environment"
+  >&2 echo "Using user $test_user for testing${normal_out}"
+else
+  test_user="root"
+  >&2 echo "${bold_out}${red_out}Warning: no suitable test user found, running tests as root${normal_out}"
+  exit 1
+fi
+
+function cleanup() {
+  # We use a loop because umount short-circuits on a failure
+  for dir in "$mount_base"/*; do
+    [[ -d "$dir" ]] && umount "$dir"
+  done
+  # kill after umount in cases of networked servers
+  kill $(cat "$root_testing_dir"/*.pid 2> /dev/null) 2> /dev/null # Only outputs errors if no PID files found
+  rm -r "$mount_base"
+  rm -r "$root_testing_dir"
+}
+
+function error() {
+  >&2 echo "${bold_out}${red_out}Error while preparing filesystem type $1, aborting${normal_out}"
+  cleanup
+  exit 1
+}
+
+while read filesystem; do
+  [[ "$filesystem" == "$root_fs" ]] && continue
+  case "$filesystem" in
+    tmpfs)
+      mkdir "$mount_base/tmpfs"
+      mount -t tmpfs tmpfs "$mount_base/tmpfs" -o "size=$testing_partition_size" || error tmpfs
+      directories+=("$mount_base/tmpfs")
+      ;;
+    ext4)
+      truncate -s "$testing_partition_size" "$root_testing_dir/ext4"
+      mkfs.ext4 -F "$root_testing_dir/ext4" || error ext4
+      mkdir "$mount_base/ext4"
+      mount -t ext4 "$root_testing_dir/ext4" "$mount_base/ext4" || error ext4
+      directories+=("$mount_base/ext4")
+      ;;
+    xfs)
+      truncate -s "$testing_partition_size" "$root_testing_dir/xfs"
+      mkfs.xfs "$root_testing_dir/xfs" || error XFS
+      mkdir "$mount_base/xfs"
+      mount -t xfs "$root_testing_dir/xfs" "$mount_base/xfs" || error XFS
+      directories+=("$mount_base/xfs")
+      ;;
+    cifs)
+      mkdir "$root_testing_dir/cifs"
+      chown nobody "$root_testing_dir/cifs" # guest == nobody user
+      smbd -D -s vagrant-tools/smb.conf
+      mkdir "$mount_base/cifs"
+      # smbd exits before startup
+      if which nc > /dev/null && [[ "$(readlink "$(which nc)")" != "ncat" ]]; then
+        while ! nc -z localhost 10445; do
+          sleep 0.5
+        done
+      else
+        while ! bash -c 'cat < /dev/null > /dev/tcp/localhost/10445' 2> /dev/null; do
+          sleep 0.5
+        done
+      fi
+      mount -t cifs //localhost/share "$mount_base/cifs" -o "port=10445,credentials=$PWD/vagrant-tools/smbcredentials" || error CIFS
+      directories+=("$mount_base/cifs")
+      ;;
+    fuse)
+      fuse_supported=true
+      ;;
+  esac
+done <<< "$filesystems"
+
+if [[ "$fuse_supported" ]]; then
+  if which sshfs > /dev/null; then
+    mkdir "$root_testing_dir/sshfs"
+    ssh-keygen -t rsa -b 4096 -C 'borgbackup@github.com' -N '' -f "$root_testing_dir/ssh_key"
```

Comment:
how long does it take to generate a 4096 bit key?
maybe we can use a shorter one to speedup the tests.